cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,316 @@
+  private static final int FETCH_SIZE = 100 * 1024;        // Use a default fetch size.
+  private static final int SO_TIMEOUT_MILLIS = 5 * 1000;           // 5 seconds.
+  private final long replicationDelayMillis;
+  private final long eventOutOfOrderMillis;
+
+  private final BrokerService brokerService;
+  private final String topic;
+  private final LoggingEventSerializer serializer;
+
+  KafkaOffsetResolver(BrokerService brokerService, KafkaPipelineConfig config) {
+    this.brokerService = brokerService;
+    this.topic = config.getTopic();
+    this.eventOutOfOrderMillis = config.getEventOutOfOrderMillis();
+    this.replicationDelayMillis = config.getReplicationDelayMillis();
+    this.serializer = new LoggingEventSerializer();
+  }
+
+  /**
+   * Check whether the message fetched with the offset {@code checkpoint.getNextOffset() - 1} contains the
+   * same timestamp as in the given checkpoint. If they match, directly return {@code checkpoint.getNextOffset()}.
+   * If they don't, search for the smallest offset of the message with the same log event time
+   * as {@code checkpoint.getNextEventTime()}
+   *
+   * @param checkpoint A {@link Checkpoint} containing the next offset of a message and its log event timestamp.
+   * @param partition the partition in the topic for searching matching offset
+   * @return the next offset of the message with smallest offset and log event time equal to
+   *         {@code checkpoint.getNextEventTime()}.
+   *         {@code -1} if no such offset can be found or {@code checkpoint.getNextOffset()} is negative.
+   */
+   long getMatchingOffset(final Checkpoint checkpoint, final int partition) {
+    // Get BrokerInfo for constructing SimpleConsumer in OffsetFinderCallback
+    BrokerInfo brokerInfo = brokerService.getLeader(topic, partition);
+    if (brokerInfo == null) {
+      return kafka.api.OffsetRequest.EarliestTime(); // Return earliest offset if brokerInfo is null
+    }
+    SimpleConsumer simpleConsumer
+      = new SimpleConsumer(brokerInfo.getHost(), brokerInfo.getPort(),
+                           SO_TIMEOUT_MILLIS, FETCH_SIZE, "simple-kafka-client");
+
+    String clientId = topic + "_" + partition;
+    // Check whether the message fetched with the offset in the given checkpoint has the timestamp from
+    // checkpoint.getNextOffset() - 1 to get the offset corresponding to the timestamp in checkpoint
+    long timeStamp = getEventTimeByOffset(simpleConsumer, partition, clientId, checkpoint.getNextOffset() - 1);
+    if (timeStamp == checkpoint.getNextEventTime()) {
+      return checkpoint.getNextOffset();
+    }
+
+    long smallestMatchingOffset = findOffsetByTime(simpleConsumer, partition, clientId, checkpoint.getNextEventTime());
+    if (smallestMatchingOffset >= 0) {
+      // Increment smallestMatchingOffset to get the next offset if it's a valid non-negative offset
+      smallestMatchingOffset++;
+    }
+    return smallestMatchingOffset;
```

Comment:
So the `findOffsetByTime` will return `-1L` if it can't the offset. In Kafka `-1L` means LatestTime, so the log processor will start polling from the latest offset? Will that result in loss of log events?